src/tools.cpp

```diff
@@ -228,16 +228,37 @@ zim::MimeCounterType zim::parseMimetypeCounter(const std::string& counterData)
 #include <unicode/translit.h>
 #include <unicode/ucnv.h>
 #include <unicode/putil.h>
+
+#define BATCH_SIZE (4*1024)
 std::string zim::removeAccents(const std::string& text)
 {
   ucnv_setDefaultName("UTF-8");
   static UErrorCode status = U_ZERO_ERROR;
   static std::unique_ptr<icu::Transliterator> removeAccentsTrans(icu::Transliterator::createInstance(
       "Lower; NFD; [:M:] remove; NFC", UTRANS_FORWARD, status));
   icu::UnicodeString ustring(text.c_str());
-  removeAccentsTrans->transliterate(ustring);
   std::string unaccentedText;
-  ustring.toUTF8String(unaccentedText);
+
+  auto nb_chars = ustring.length();
+  if (nb_chars <= BATCH_SIZE) {
+    // Remove accents in one step.
+    removeAccentsTrans->transliterate(ustring);
+    ustring.toUTF8String(unaccentedText);
+  } else {
+    auto current_pos = 0;
+    icu::UnicodeString current_ustring;
+    while (current_pos < nb_chars) {
+      // Remove accents by batch of BATCH_SIZE "chars" to avoid working on
+      // a too long string and spending to much time memcpy things.
+      auto end = ustring.getChar32Limit(current_pos+BATCH_SIZE);
```

Comment:
The last iteration will most likely overflow here and [`getChar32Limit()`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeString.html#a3757a05c72d31543eb47c6af906949c0) requires its argument to be a valid offset. The added unit test fails to catch this bug since the size of the test data is an exact multiple of the batch size.

Reply:
In the detailed description of UnicodeString (https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeString.html#details) it is said that "If indexes are out of bounds for a UnicodeString object (< 0 or > length()) then they are "pinned" to the nearest boundary."
Se we should have to handle out of bound here. (But I will adapt the test anyway to be sure)